nixos/sftpgo: init, nixos/tests/sftpgo: init

yayayayaka opened the pull request:
SFTPGo is a fully featured and highly configurable SFTP server with optional HTTP/S, FTP/S and WebDAV support.

https://github.com/drakkan/sftpgo

###### Description of changes
- adds an sftpgo NixOS module
- adds a NixOS test for sftpgo
<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.11 Release Notes (or backporting 22.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2211-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

yu-re-ka commented on nixos/tests/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,378 @@
+          ensurePermissions."DATABASE sftpgo" = "ALL PRIVILEGES";
+        }];
+        # Do not listen on TCP if already using Unix sockets
+        settings.listen_addresses = mkForce "";
```
.

pennae commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,370 @@
+        ReadWritePaths = cfg.dataDir;
+        LimitNOFILE = 8192; # taken from upstream
+        KillMode = "mixed";
+        ExecStart = "${cfg.package}/bin/sftpgo serve ${escapeShellArgs cfg.extraArgs}";
```
`Exec*` aren't actually shell, `utils.escapeSystemdExecArgs` is better suited

yayayayaka replied:
Good to know, thanks!

pennae commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,370 @@
+      after = [ "network.target" ];
+      wantedBy = [ "multi-user.target" ];
+
+      environment =
+        recursiveUpdate (mapAttrs (name: mkDefault) {
+          SFTPGO_CONFIG_FILE = configFile;
+          SFTPGO_LOG_FILE_PATH = ""; # log to journal
+        }) (optionalAttrs (cfg.loadDataFile != null) {
+          SFTPGO_LOADDATA_FROM = cfg.loadDataFile;
+        });
```
```suggestion
      environment = {
        SFTPGO_CONFIG_FILE = mkDefault configFile;
        SFTPGO_LOG_FILE_PATH = mkDefault ""; # log to journal
        SFTPGO_LOADDATA_FROM = mkIf (cfg.loadDataFile != null) cfg.loadDataFile;
      };
```
is this viable?

yayayayaka replied:
No complaints when running the test, seems fine.

pennae commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,370 @@
+in
+{
+  options.services.sftpgo = {
+    enable = mkEnableOption "Enable sftpgo";
```
```suggestion
    enable = mkEnableOption "sftpgo";
```

pennae commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,370 @@
+
+    dataDir = mkOption {
+      type = types.str;
+      default = "/var/lib/sftpgo";
```
you differentiate `dataDir` and `StateDirectory`, but in the default setting they'll be the same. would it make sense to make this a `nullOr` and bind-mount only when set?

yayayayaka replied:
Take a look at 7df8471b63c. Maybe this might be a solutoin?

pennae replied:
yeah, that looks reasonable too :)

SuperSandro2000 commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,370 @@
+  defaultUser = "sftpgo";
+  settingsFormat = pkgs.formats.json {};
+  configFile = settingsFormat.generate "sftpgo.json" cfg.settings;
+  hasPrivilegedPorts = builtins.any (port: port > 0 && port < 1024) (
+    builtins.catAttrs "port" (cfg.settings.httpd.bindings
```
If there are lib counterparts for builtins those functions should be used

SuperSandro2000 commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,370 @@
+      type = types.package;
+      default = pkgs.sftpgo;
+      defaultText = literalExpression "pkgs.sftpgo";
+      description = mdDoc ''
```
```suggestion
      description = ''
```

SuperSandro2000 commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,370 @@
+    extraArgs = mkOption {
+      type = with types; listOf str;
+      default = [];
+      description = mdDoc ''
```
```suggestion
      description = ''
```

pennae replied:
please don't we plan on making `mdDoc` mandatory for a transitional period to ensure that all docs are converted correctly

SuperSandro2000 replied:
There is currently no documentation about this which makes it really hard to enforce in PRs and it probably should require a grace period over one release. Also there will definitely be PRs that miss this.

pennae replied:
our plan was to add that rule to the documentation once all options were converted, add a nice error message to the manual build for cases where it was forgotten, and keep that in place for a major. if things work out 22.11 would release with mdDoc only, 23.05 would remove it again. in unstable we'd want at least a month between adding enforcement and dropping it again, and considering that we still have no agreement on which markdown processor to use for the manual that doesn't seem like a very tight schedule.

SuperSandro2000 replied:
If we want to start changing this in review process the documentation must be in place first. Otherwise I have nothing to refer people to.

pennae replied:
we already have a note in the documentation that writing new docs as docbook is discouraged: https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/development/option-declarations.section.md?plain=1#L58-L66. there's no note yet about docbook being disallowed because there's so much docbook left in nixpkgs. if you think there should be more or clearer docs please do tell, this should be communicated as clearly as possible.

SuperSandro2000 commented on nixos/tests/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,376 @@
+          };
+        in {
+          # SFTPGo admin user
+          admin = recursiveUpdate commonAttrs {
```
This could be safely swapped with ``//``

SuperSandro2000 commented on pkgs/servers/web-apps/sftpgo/default.nix:
```diff
@@ -0,0 +1,33 @@
+{ lib
+, buildGo118Module
+, fetchFromGitHub
+}:
+
+buildGo118Module rec {
```
```suggestion
, buildGoModule
, fetchFromGitHub
}:

buildGoModule rec {
```

pennae commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,368 @@
+      description = mdDoc ''
+        Additional command line arguments to pass to the sftpgo daemon.
+      '';
+      example = [ "--log-level info" ];
```
this probably won't work when escaped (or does sftpgo split arguments on its own too?)

pennae commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,368 @@
+      type = with types; nullOr path;
+      description = mdDoc ''
+        Path to a json file containing users and folders to load (or update) on startup.
+        Check the [documentation](https://github.com/drakkan/sftpgo/blob/main/docs/full-configuration.md) for the `--loaddata-from` command line argument for more info.
```
```suggestion
        Check the [documentation](https://github.com/drakkan/sftpgo/blob/main/docs/full-configuration.md)
        for the `--loaddata-from` command line argument for more info.
```

pennae commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,368 @@
+                  default = "127.0.0.1";
+                  description = mdDoc ''
+                    Network listen address. Leave blank to listen on all available network interfaces.
+                    On *NIX you can specify an absolute path to listen on a Unix-domain socket
```
```suggestion
                    On *NIX you can specify an absolute path to listen on a Unix-domain socket.
```
the other address options seem to be missing their `.` too

pennae commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,368 @@
+                  default = 1;
+                  description = mdDoc ''
+                    Encryption scheme:
+                    0: No encryption
+                    1: TLS
+                    2: STARTTLS
```
might be more readable as a list if we can't use a more verbose type:
```suggestion
                    - `0`: No encryption
                    - `1`: TLS
                    - `2`: STARTTLS
```

SuperSandro2000 commented on pkgs/servers/web-apps/sftpgo/default.nix:
```diff
@@ -0,0 +1,33 @@
+  subPackages = [ "." ];
+
+  postInstall = ''
+    cp -r ./{openapi,static,templates} $out/
```
We shouldn't put those at ``/``. Maybe $out/share/sftpgo/* would be a better place?

yayayayaka replied:
Changed to `$out/share/sftpgo/`

SuperSandro2000 commented on pkgs/servers/web-apps/sftpgo/default.nix:
```diff
@@ -0,0 +1,33 @@
+  '';
+
+  meta = with lib; {
+    description = ''
+      A Fully featured and highly configurable SFTP server with optional HTTP/S,
+      FTP/S and WebDAV support
+    '';
```
```suggestion
    description = "Fully featured and highly configurable SFTP server with optional HTTP/S, FTP/S and WebDAV support";
```

aanderse commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,377 @@
+      example = [ "--log-level" "info" ];
+    };
+
+    dataDir = mkOption {
```
You use 3 different methods to create this directory: `StateDirectory`, `tmpfiles`, and `createHome`.

You should not use `createHome`. In my opinion you should use `StateDirectory`.

yu-re-ka replied:
`StateDirectory` only works if it is inside `/var/lib`.

I would argue that the dataDir should not be configurable.

aanderse replied:
> `StateDirectory` only works if it is inside `/var/lib`.

Indeed. A good pattern in NixOS is to leave the user with an escape hatch that we don't provision directories for them. When dealing with network filesystems, etc... this can cause users problems. See an example of what is a good way to handle state directory creation here: https://github.com/NixOS/nixpkgs/blob/de5448dab588ad41aef40f8c7c0c230981656698/nixos/modules/services/security/cfssl.nix#L11


> maybe do `createHome` only if the dataDir is _not_ `/var/lib/sftpgo`?

Never use `createHome` to create state directories because it ends poorly for some users needs (see prior comment about network mounts).

> Some might argue that the dataDir should not be configurable in the first place, and it is the user's task to somehow bind-mound or symlink the data to /var/lib.

This certainly would make life easier for developers... though at the cost of the end user. Conditionally creating the directory as in the example I linked above is the best compromise on my opinion.

SuperSandro2000 commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,377 @@
+          ExecReload = "${pkgs.util-linux}/bin/kill -s HUP $MAINPID";
+
+          # Service hardening
+          CapabilityBoundingSet = optional hasPrivilegedPorts "CAP_NET_BIND_SERVICE";
```
We can simplify this a lot by just always allowing this, not like it is a high security risk.

SuperSandro2000 commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,377 @@
+          RestrictRealtime = true;
+          RestrictSUIDSGID = true;
+          SystemCallArchitectures = "native";
+          SystemCallFilter = "~@clock @cpu-emulation @debug @mount @obsolete @reboot @swap @privileged @resources";
```
Isn't @resources crashing with go binaries compiled with 1.19?

SuperSandro2000 commented on nixos/modules/services/web-apps/sftpgo.nix:
```diff
@@ -0,0 +1,377 @@
+          SystemCallArchitectures = "native";
+          SystemCallFilter = "~@clock @cpu-emulation @debug @mount @obsolete @reboot @swap @privileged @resources";
+          UMask = "0077";
+        } // optionalAttrs hasPrivilegedPorts {
+          AmbientCapabilities = "CAP_NET_BIND_SERVICE";
+        })
+        (mkIf (cfg.dataDir == "/var/lib/sftpgo") {
+          StateDirectory = "sftpgo";
+        })
```
Why not use either mkMerge or optionalAttrs here?

yayayayaka replied:
Updated. Looking better now?

yayayayaka committed:
nixos/sftpgo: init

A fully featured and highly configurable SFTP server with optional
HTTP/S, FTP/S and WebDAV support.

https://github.com/drakkan/sftpgo

yayayayaka committed:
nixos/tests/sftpgo: init

yayayayaka committed:
sftpgo: Include openapi, static and templates folders

Those are needed for serving the web client

yayayayaka committed:
sftpgo: add yayayayaka to maintainers

yayayayaka committed:
nixos/doc: add release note for sftpgo

SuperSandro2000 reviewed (approved):
I skipped mostly over the test but should be alright
